### ISSUE TITLE
Acronym tooltip disambiguation

### ISSUE THREAD

**Issue by arvindpdmn** [DevopediaOrg/webapp#468]
Suppose an article uses the acronym "CSS". On mouse hover, the expanded form is shown as a tooltip. If there are multiple expansions (such as `Cascading Style Sheets` and `Chirp Spread Spectrum`), all the expansions are shown. This is a limitation of Devopedia platform. Instead, only the most relevant expansion should be shown. In the article titled [CSS Grid Layout](https://devopedia.org/css-grid-layout), the only sensible expansion of CSS is `Cascading Style Sheets`.

The job of the NLP model is to look at the current article context or even paragraph context to know which is the correct expansion.

Where do the expansions come from? Devopedia maintains acronyms in two database tables:
- short form name and expansion of each acronym
- a mapping to one or more articles in which the acronym is defined

In other words, an acronym and it's expansion doesn't enter the Devopedia system until it's defined in some article. All article content plus the two acronym tables will be shared as data for this task.

Optional: The scope of this work can be expanded to acronyms that are not in the system. In these case, the expansions would have to come from external sources (databases, Wikipedia, Google search, etc.) This means that if a new acronym is used by an author without defining, we'll still be able to show the correct expansion.

**Comment by arvindpdmn** [comment]
### Examples Extracted by Devopedia

- Bipolar Loud Male (BiLM) that includes multiple letters from a word.
- It was name it Made Sense (IMS) framework where actual word starts with lowercase.
- This is the Under Framework (TUF).
- This of Object-oriental Programming (OOP) excludes "of" in the acronym.
- Original Object-oriented Pill (OOP) includes "Original" in the acronym.
- This is an acronym without hyphen: object oriented problem (OOP).
- Test Positive (TP) is a valid acronym whereas Test Positives (TP) is ignored.
- Test Negatives (TNs) is a valid plural form of acronym.
- Recursive with quote: GNL's Not Linux (GNL).
- Recursive: OOP Object-oriented Problem (OOP).
- Intranet-Drafts (I-Ds) is a valid acronyms with plural form and hyphen.
- In databases, we have Basically available, Soft-state, and Eventually consistent (BASE) that is mixed case with commas and ignore stopwords.
- Reversal of expansion and acronym such as ACID (Amply, Consistency, Isolation, and Durability).
- Amazon Web Services (AWS).
- Carrier Sense Multiple Access/Collision Detection (CSMA/CD) includes non-alphabets in acronym.
- Carrier-Sense Multiple Access/Collision Free (CSMA/CF) includes non-alphabets in acronym and hyphen in expansion.
- extended DRX (eDRX) in which expansion and acronym starts with lowercase.
- Generative Pre-trained Transformer (GPT).
- Global Vectors (GloVe) in which acronym is mixed case.
- Head-Driven Phrase Structure Grammar (HPSG).
- Hierarchical SBM (hSBM) which is a case of sub-expansion.
- LinuX Containers (LXC) which is a case of uppercase occurring in the middle.
- Multi Document Summarization (MDS).
- Wi-Fi Multimedia (WMM) where a word is split into multiple words.
- Multidimensional Scaling (MDS) and Multi Dimensional Scaling (MDS) are equivalent.
- Nominating Committee (NomCom) which is the case of using starting substrings of words.
- Open Data Commons Open Database License (ODC-ODbL) where acronym has hyphen and multiple letters from a word.
- Out of Vocabulary (OOV).
- Quality-as-a-Service (QaaS) where expansion has hyphens.
- Python Package Index (PyPI) where acronym has multiple letters from a word.
- Stanford Question Answering Dataset (SQuAD).
- Teacher Free Knowledge Distillation (Tf-KD) where start of some words become lowercase in acronym.
- Virtual Dynamic Shared Object (vDSO).
- Voice over Wi-Fi (VoWi-Fi).
- Wireless Gigabit (WiGig).
- ZAP as a Service (ZaaS) and ZAP-as-a-Service (ZaaS) are equivalent.
- Distributed-File/System (DFS).


### Examples Not Extracted by Devopedia

- Acronyms starting with multiple lowercase letters from a word such as bidirectional language model (biLM). But we can extract bidirectional language model (BiLM).
- Non-common "compound" words (which could be fixed by looking for sub-words):
  - Packet Gateway (PGW)
  - enhanced Mobile Broadband (eMBB)
  - OpenID Connect (OIDC)
  - Natural Language Toolkit (NLTK)
- Word to digit:
  - Lightweight Machine to Machine (LWM2M)
  - Fifth Generation (5G)
- Abbreviations: Reset (RST), Acknowledgement (ACK)
- Multi-letters (but Representational State Transfer (ReST) is captured):
  - Representational State Transfer (REST)
  - REpresentational State Transfer (REST)
  - Well-Architected Framework (WAFR)
  - Internet Society (ISOC)
- Different Order:
  - Total Sum of Squares (SST)
  - Web Ontology Language (OWL)
- Extra Words:
  - Corpus of Contemporary American English (COCA)
  - Lancaster-Oslo/Beren Corpus (LOB)
  - Speech API Phone Set (SAPI)
- Combinations:
  - Signal-to-Noise Ratio (SNR) that combines hypen + stopword
  - eXtensible Messaging and Presence Protocol (XMPP) that combines midcaps + stopword
  - Connected Hospital over IP (CHIP) that combine sub-expansion + stopword


### Other Issues
- Hierarchical SBM (hSBM) and Hierarchical SBM (H-SBM) are treated differently since acronyms have different forms.
- Distribution Center (DC) and Distribution Centre (DC) are treated as two different acronyms
- When displayed, MR-DC is expanded separately for MR and DC but not as a composite.

**Comment by arvindpdmn** [comment]
Adeft links:
- Paper on Adeft: https://joss.theoj.org/papers/10.21105/joss.01708
- Code also available: https://github.com/indralab/adeft
- Docs: https://adeft.readthedocs.io/en/latest/modules/index.html

Oracle links:
- Paper: https://github.com/maya124/AcronymLookup/blob/master/Machine-Learning%20Approach%20for%20Cross-Domain%20Acronym%20Definition%20Identification.pdf
- Code: https://github.com/maya124/AcronymLookup

Other links:
- https://acronomy.lw1.at/
- https://github.com/Findus23/acronomy
- https://cloud.google.com/architecture/building-real-time-embeddings-similarity-matching-system

Ideas:
- BOW / N-gram
- Word2vec
- Topic modelling / LDA
- SVM

**Comment by arvindpdmn** [comment]
Our code: https://github.com/teja0508/AcronymLookup
A related issue: update article caches when an acronym changes. Eg. A new acronym is used in article A. Later article B defines it for the first time. Article cache of A must be updated.

**Comment by arvindpdmn** [comment]
Noting a couple of issues (which can be added to test data):
- Double: `[JS]([javascript](javascript))` files
      : Bidirectional Recurrent Neural Network `([BRNN](bidirectional-rnn))`
- Wrong : Chirp Spread Spectrum (`[CSS](cascading-style-sheets)`) modulation.
      : `AMF (Additive Manufacturing File)` but incorrect in 5G context

**Comment by arvindpdmn** [comment]
Disambiguation implemented as a web service with GraphQL. An incomplete design is below:
```
+ Add to cron (20hour UTC):
exec("cd ../tools/kaggle; $phpExec main.php");
+ Acronyms:
CREATE TABLE `xxxxx_acronyms_disambig` (
  `version_id` int(10) UNSIGNED NOT NULL,
  `position` tinyint UNSIGNED NOT NULL,
  `acronyms_id` int(10) UNSIGNED NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
ALTER TABLE `xxxxx_acronyms_disambig`
  ADD UNIQUE KEY `idx_version_id_position` (`version_id`, `position`);

OK
{"query":"{ acronym {name,description} }"}
{"query":"{ acronym(id:200) {name,description} }"}
{"query":"{ createAcronym(input: {name:\"ADC\", description:\"Ab Dase Cls\"}) }"}

{"query":"{ hello(id:200) {name,description} }"}
{"query":"{ createAcr(name:\"xxx\", description:\"yyy\") {id} }"}
{"query":"{ createAcr(name:\"ABC\", description:\"Ab Base Cls\") }"} # return plain Int
{"query":"{ createAcr(input: {name:\"ADC\", description:\"Ab Dase Cls\"}) }"}

KO
{"query":"{ createAcr(input:{id:440}) {id} }"}
{"mutation":"{ createAcr({id: 400, name: 'xxx', description: 'yyy'}) {Acronym {id, name}} }"}

type Query {
    hello(id: Int!): [Acronym]
    createAcr(name: String, description: String): Int!
}

type Acronym {
    id: Int!
    name: String!
    description: String!
}

schema {
  query: Query
  mutation: Mutation
}

type Query {
    hello: [Acronym]
}

type Acronym {
    id: Int!
    name: String!
    description: String!
}

type Mutation {
    createAcr(input: Acronym): Acronym
    updateAcr(id: Int!, input: Acronym): Acronym
}
```
